src/render/src/handle.rs

```diff
@@ -0,0 +1,85 @@
+use std::marker::PhantomData;
+use std::sync::{Arc, mpsc};
+use std::{cmp, hash};
+
+use core;
+use Backend;
+
+pub(crate) enum Garbage<B: Backend> {
+    ShaderLib(B::ShaderLib),
+    Buffer(B::Buffer),
+    Image(B::Image),
+    RenderTargetView(B::RenderTargetView),
+    DepthStencilView(B::DepthStencilView),
+    ConstantBufferView(B::ConstantBufferView),
+    ShaderResourceView(B::ShaderResourceView),
+    UnorderedAccessView(B::UnorderedAccessView),
+    Sampler(B::Sampler),
+}
+
+pub(crate) type GarbageSender<B> = mpsc::Sender<Garbage<B>>;
+pub(crate) type GarbageReceiver<B> = mpsc::Receiver<Garbage<B>>;
+pub(crate) fn garbage_channel<B: Backend>() -> (GarbageSender<B>, GarbageReceiver<B>) {
+    mpsc::channel()
+}
+
+macro_rules! resource_handle {
+    ($handle:ident, $guard:ident, $resource:ident) => {
+        #[derive(Clone, Debug)]
+        struct $guard<B: Backend>(Option<B::$resource>, GarbageSender<B>);
```

Comment:
Would you mind explaining why we need `Option<_>` here?

Reply:
I need ownership on `Drop` to send the resource on the garbage channel for future cleanup.